src/node/coinstats.cpp

```diff
@@ -193,4 +187,44 @@ static void FinalizeHash(MuHash3072& muhash, CCoinsStats& stats)
     stats.hashSerialized = out;
 }
 static void FinalizeHash(std::nullptr_t, CCoinsStats& stats) {}
+
+std::optional<CCoinsStats> LookupUTXOStatsWithIndex(CoinStatsIndex& coin_stats_index, const CBlockIndex* pindex)
+{
+    CCoinsStats stats{Assert(pindex)->nHeight, pindex->GetBlockHash()};
+
+    stats.index_used = true;
+    if (!coin_stats_index.LookUpStats(pindex, stats)) {
+        return std::nullopt;
+    }
+
+    return stats;
+}
+
+std::optional<CCoinsStats> LookupUTXOStatsWithIndex(CoinStatsIndex& coin_stats_index, CCoinsView* view, BlockManager& blockman)
+{
+    CBlockIndex* pindex = WITH_LOCK(::cs_main, return blockman.LookupBlockIndex(view->GetBestBlock()));
+
+    return LookupUTXOStatsWithIndex(coin_stats_index, pindex);
+}
+
+std::optional<CCoinsStats> GetUTXOStats(CCoinsView* view, BlockManager& blockman, CoinStatsHashType hash_type, const std::function<void()>& interruption_point, const CBlockIndex* pindex, bool index_requested)
+{
+    // Use CoinStatsIndex if it is requested and available and a hash_type of Muhash or None was requested
+    if ((hash_type == CoinStatsHashType::MUHASH || hash_type == CoinStatsHashType::NONE) && g_coin_stats_index && index_requested) {
+        if (pindex) {
+            return LookupUTXOStatsWithIndex(*g_coin_stats_index, pindex);
+        } else {
+            return LookupUTXOStatsWithIndex(*g_coin_stats_index, view, blockman);
+        }
+    }
+
+    // pindex is ignored if the coinstats index is not available or wasn't
+    // requested. Coinstats for the view are returned instead.
+    //
+    // This is ok because we never call GetUTXOStats in a way where we can't use
+    // the coinstatsindex and the specified pindex is not also the view's best
+    // block.
```

Comment:
In commit "coinstats: Separate hasher/index lookup codepaths" (99b7dced926eeba1ac0b181feb18ebb7ce35504e)

Seems like it would be easy for code calling this function to be unaware of this assumption. Could check it with
`assert(!pindex || pindex->GetBlockHash() == view->GetBestBlock());`

Reply:
Right! Fixed in 1352e410a5b84070279ff28399083cb3ab278593